### PR TITLE
feat(python): Support Formatting via ruff

### PR DESCRIPTION
**Pull request by noanflaherty**
This PR introduces code formatting via ruff and updates tests accordingly.

### CODE REVIEW

File: generators/pythonv2/codegen/src/ast/PythonFile.ts [hunk 1]
```diff
@@ -20,7 +20,7 @@ export class PythonFile extends AstNode {
     public readonly moduleName: string;
     public readonly path: string[];
     public readonly name: string;
-    private readonly statements: (Class | Method | CodeBlock)[] = [];
+    private readonly statements: AstNode[] = [];
```
Comment (noanflaherty) [inline generators/pythonv2/codegen/src/ast/PythonFile.ts:23]: Unrelated change I'm sneaking in. You should be able to add any sort of node as a statement to a file imo.

File: generators/pythonv2/codegen/src/ast/__test__/Class.test.ts [hunk 2]
```diff
@@ -1,11 +1,19 @@
 import { python } from "../..";
+import { Writer } from "../core/Writer";
 
 describe("class", () => {
+    let writer: Writer;
+
+    beforeEach(() => {
+        writer = new Writer();
+    });
```
Comment (noanflaherty) [inline generators/pythonv2/codegen/src/ast/__test__/Class.test.ts:9]: You'll find that I've updated all tests to call the writer's `toStringFormatted` method rather than adding a `toStringFormatted` method to the `AstNode`.

I describe why below.

File: generators/pythonv2/codegen/src/ast/core/Writer.ts [hunk 3]
```diff
@@ -14,6 +15,11 @@ export class Writer extends AbstractWriter {
         return this.buffer;
     }
 
+    public async toStringFormatted(): Promise<string> {
+        await init();
+        return format(this.buffer);
+    }
```
Comment (noanflaherty) [inline generators/pythonv2/codegen/src/ast/core/Writer.ts:21]: I intentionally chose to add this to the writer rather than the AstNode. This is because the writer has context needed for codegen that the AstNode doesn't have – primarily, references.

This was important for `PythonFile`, which relies on the fact that the writer has accumulated the references needed to generate imports for the file.

Comment (noanflaherty) [reply generators/pythonv2/codegen/src/ast/core/Writer.ts:21]: Side Note: I wonder if there's a way to use a global to ensure that in a given process, we only ever call `await init();` once? I noticed tests were a bit slower and I suspect this is why.

File: generators/pythonv2/codegen/src/ast/PythonFile.ts (hunk 1 above)
Comment (dsinghvi) [reply generators/pythonv2/codegen/src/ast/PythonFile.ts:23]: nice

File: generators/pythonv2/codegen/src/ast/core/Writer.ts (hunk 3 above)
Comment (dsinghvi) [reply generators/pythonv2/codegen/src/ast/core/Writer.ts:21]:
```
export abstract class AstNode extends AbstractAstNode {
    /**
     * Writes the node to a string.
     */
    public toString(): string {
        const writer = new Writer();
        this.write(writer);
        return writer.toString();
    }

    public toStringFormatted(): string {
        return prettier.format(this.toString(), { parser: "typescript", tabWidth: 4, printWidth: 120 });
    }
}
```

Comment (dsinghvi) [reply generators/pythonv2/codegen/src/ast/core/Writer.ts:21]: this is how i do it in typescript

**Review by dsinghvi — APPROVED** [review]
python guru 🚀
